File: packages/protocol/src/transfer-receiver-resume.test.ts
import { describe, it, expect } from 'vitest';
import { createHash } from 'node:crypto';
import { deriveTransferKeys } from './keyschedule.js';
import { open, seal, type FrameHeaderInput } from './aead.js';
import { FrameType, type Manifest } from './transfer.js';
import { FRAME_VERSION } from './constants.js';
import { decodeControl, encodeControl } from './transfer-messages.js';
import { bytesToHex } from './bytes.js';
import { sha256 } from './webcrypto.js';
import { MemorySink, type Digest } from './transfer-ports.js';
import { TransferReceiver, type ReceiverResumeState } from './transfer-receiver.js';
import { manifestFingerprint } from './journal.js';

function nodeDigest(): Digest {
  const h = createHash('sha256');
  return { update: (b) => void h.update(b), hexDigest: () => h.digest('hex') };
}
const master = new Uint8Array(32).fill(9);
const ID = 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa';

async function build(keys: Awaited<ReturnType<typeof deriveTransferKeys>>) {
  let ctr = 0;
  const frames: Uint8Array[] = [];
  const push = async (header: FrameHeaderInput, payload: Uint8Array) => {
    frames.push(await seal(keys.o2j, ctr++, header, payload));
  };
  const ctrl = (type: FrameType, msg: Parameters<typeof encodeControl>[0]) =>
    push(
      { version: FRAME_VERSION, type, flags: 0, fileIdx: 0, blockIdx: 0, frameOff: 0 },
      encodeControl(msg),
    );
  return { frames, push, ctrl };
}

async function expectedManifest(): Promise<Manifest> {
  return {
    type: FrameType.Manifest,
    transferId: ID,
    files: [
      {
        idx: 0,
        name: 'f',
        size: 20,
        mime: '',
        lastModified: 0,
        blockSize: 8,
        blocks: 3,
        fileDigest: createHash('sha256').update(new Uint8Array(20)).digest('hex'),
      },
    ],
    totalSize: 20,
  };
}

async function streamFile(b: Awaited<ReturnType<typeof build>>, data: Uint8Array): Promise<void> {
  for (let blk = 0; blk * 8 < data.length; blk++) {
    const start = blk * 8;
    const end = Math.min(start + 8, data.length);
    const block = data.subarray(start, end);
    for (let off = 0; off < block.length; off += 4) {
      const frag = block.subarray(off, Math.min(off + 4, block.length));
      const last = off + frag.length === block.length;
      await b.push(
        {
          version: FRAME_VERSION,
          type: FrameType.BlockData,
          flags: last ? 1 : 0,
          fileIdx: 0,
          blockIdx: blk,
          frameOff: off,
        },
        frag,
      );
    }
    await b.ctrl(FrameType.BlockHash, {
      type: FrameType.BlockHash,
      fileIdx: 0,
      blockIdx: blk,
      sha256: bytesToHex(await sha256(block)),
    });
  }
}

async function runReceiver(
  manifest: Manifest,
  stream: boolean,
  resume: ReceiverResumeState | undefined,
  complete = true,
): Promise<{
  result: Awaited<ReturnType<TransferReceiver['done']['then']>> | undefined;
  err: unknown;
  back: Parameters<typeof decodeControl>[0][];
}> {
  const keys = await deriveTransferKeys(master);
  const b = await build(keys);
  await b.ctrl(FrameType.Manifest, manifest);
  if (stream) {
    await streamFile(b, new Uint8Array(20));
  }
  if (complete) {
    await b.ctrl(FrameType.Complete, {
      type: FrameType.Complete,
      fileDigest: manifest.files[0]!.fileDigest,
    });
  }
  const sink = new MemorySink();
  const backChannel: Uint8Array[] = [];
  const receiver = new TransferReceiver({
    send: (f) => void backChannel.push(f),
    sendDir: keys.j2o,
    recvDir: keys.o2j,
    sendCounterStart: 0,
    recvCounterStart: 0,
    createDigest: nodeDigest,
    sink,
    resume,
  });
  for (const f of b.frames) receiver.handle(f);
  let result: Awaited<ReturnType<TransferReceiver['done']['then']>> | undefined;
  let err: unknown;
  try {
    result = await receiver.done;
  } catch (e) {
    err = e;
  }
  const back: Parameters<typeof decodeControl>[0][] = [];
  let c = 0;
  for (const f of backChannel) {
    const o = await open(keys.j2o, c++, f);
    back.push(decodeControl(o.plaintext));
  }
  return { result, err, back };
}

describe('TransferReceiver resume validation (V13-PR06)', () => {
  it('fails closed on bad resume seeds without advertising any resume_state', async () => {
    const manifest = await expectedManifest();
    const fp = await manifestFingerprint(manifest);
    const base = (mutate: (r: ReceiverResumeState) => void): ReceiverResumeState => {
      const r: ReceiverResumeState = {
        transferId: ID,
        manifestFingerprint: fp,
        files: new Map([[0, { haveBlocks: 0, seedDigest: nodeDigest() }]]),
      };
      mutate(r);
      return r;
    };
    const cases: { name: string; resume: ReceiverResumeState; want: RegExp }[] = [
      {
        name: 'uppercase fingerprint',
        resume: base((r) => (r.manifestFingerprint = r.manifestFingerprint.toUpperCase())),
        want: /resume seed manifestFingerprint must be 64 lowercase hex characters/,
      },
      {
        name: 'short fingerprint',
        resume: base((r) => (r.manifestFingerprint = fp.slice(0, 10))),
        want: /resume seed manifestFingerprint must be 64 lowercase hex characters/,
      },
      {
        name: 'wrong fingerprint',
        resume: base((r) => (r.manifestFingerprint = '0'.repeat(64))),
        want: /resume seed manifest fingerprint does not match the authenticated manifest/,
      },
      {
        name: 'empty file set',
        resume: base((r) => (r.files = new Map())),
        want: /resume seed covers 0 files, manifest has 1/,
      },
      {
        name: 'haveBlocks above blocks',
        resume: base(
          (r) => (r.files = new Map([[0, { haveBlocks: 4, seedDigest: nodeDigest() }]])),
        ),
        want: /resume seed haveBlocks 4 out of range for file 0 \(blocks 3\)/,
      },
    ];
    for (const c of cases) {
      const { err, back } = await runReceiver(manifest, false, c.resume);
      expect(err, c.name).toBeInstanceOf(Error);
      expect(String((err as Error).message), c.name).toMatch(c.want);
      expect((err as Error).message).toMatch(/sink_error/);
      // The only back-channel frame is the terminal fail; no resume_state was advertised.
      expect(back.length, c.name).toBe(1);
      expect(back[0]!.type, c.name).toBe(FrameType.Fail);
    }
  });

  it('advertises a fingerprint-bound resume_state and streams only the missing blocks', async () => {
    const manifest = await expectedManifest();
    const fp = await manifestFingerprint(manifest);
    const prefix = new Uint8Array(16);
    const seed = nodeDigest();
    seed.update(prefix);
    const { err, back } = await runReceiver(manifest, true, {
      transferId: ID,
      manifestFingerprint: fp,
      files: new Map([[0, { haveBlocks: 2, seedDigest: seed }]]),
    });
    expect(err).toBeUndefined();
    expect(back.length).toBe(5); // resume_state, ack ×3, done
    const rs = back[0] as Extract<(typeof back)[number], { type: FrameType.ResumeState }>;
    expect(rs.type).toBe(FrameType.ResumeState);
    expect(rs.manifestFingerprint).toBe(fp);
    expect(rs.files).toEqual([{ idx: 0, haveBlocks: 2 }]);
    expect(back[back.length - 1]!.type).toBe(FrameType.Done);
  });

  it('ignores a resume seed for a different transfer (fresh all-zero resume_state)', async () => {
    const manifest = await expectedManifest();
    const staleSeed = nodeDigest();
    staleSeed.update(new Uint8Array(8));
    const { err, back } = await runReceiver(manifest, true, {
      transferId: 'b'.repeat(32),
      manifestFingerprint: '0'.repeat(64),
      files: new Map([[0, { haveBlocks: 2, seedDigest: staleSeed }]]),
    });
    expect(err).toBeUndefined();
    const rs = back[0] as Extract<(typeof back)[number], { type: FrameType.ResumeState }>;
    expect(rs.files).toEqual([{ idx: 0, haveBlocks: 0 }]);
  });

  it('re-answers an identical resume_state on a duplicate manifest and rejects a different one', async () => {
    const manifest = await expectedManifest();
    const fp = await manifestFingerprint(manifest);
    const seed = nodeDigest();
    seed.update(new Uint8Array(8));

    // Identical duplicate manifest → identical resume_state both times.
    {
      const keys = await deriveTransferKeys(master);
      const b = await build(keys);
      await b.ctrl(FrameType.Manifest, manifest);
      await b.ctrl(FrameType.Manifest, manifest);
      await streamFile(b, new Uint8Array(20));
      await b.ctrl(FrameType.Complete, {
        type: FrameType.Complete,
        fileDigest: manifest.files[0]!.fileDigest,
      });
      const sink = new MemorySink();
      const backChannel: Uint8Array[] = [];
      const receiver = new TransferReceiver({
        send: (f) => void backChannel.push(f),
        sendDir: keys.j2o,
        recvDir: keys.o2j,
        sendCounterStart: 0,
        recvCounterStart: 0,
        createDigest: nodeDigest,
        sink,
        resume: {
          transferId: ID,
          manifestFingerprint: fp,
          files: new Map([[0, { haveBlocks: 1, seedDigest: seed }]]),
        },
      });
      for (const f of b.frames) receiver.handle(f);
      await receiver.done;
      const backs: Parameters<typeof decodeControl>[0][] = [];
      let c = 0;
      for (const f of backChannel)
        backs.push(decodeControl((await open(keys.j2o, c++, f)).plaintext));
      const first = backs[0] as Extract<(typeof backs)[number], { type: FrameType.ResumeState }>;
      const second = backs[1] as Extract<(typeof backs)[number], { type: FrameType.ResumeState }>;
      expect(first.type).toBe(FrameType.ResumeState);
      expect(second).toEqual(first);
    }

    // A different manifest after the first was applied is a protocol violation.
    {
      const other: Manifest = {
        ...manifest,
        files: [
          {
            ...manifest.files[0]!,
            size: 40,
            blocks: 5,
            fileDigest: createHash('sha256').update(new Uint8Array(40)).digest('hex'),
          },
        ],
        totalSize: 40,
      };
      const keys = await deriveTransferKeys(master);
      const b = await build(keys);
      await b.ctrl(FrameType.Manifest, manifest);
      await b.ctrl(FrameType.Manifest, other);
      const sink = new MemorySink();
      const backChannel: Uint8Array[] = [];
      const receiver = new TransferReceiver({
        send: (f) => void backChannel.push(f),
        sendDir: keys.j2o,
        recvDir: keys.o2j,
        sendCounterStart: 0,
        recvCounterStart: 0,
        createDigest: nodeDigest,
        sink,
        resume: {
          transferId: ID,
          manifestFingerprint: fp,
          files: new Map([[0, { haveBlocks: 1, seedDigest: seed }]]),
        },
      });
      for (const f of b.frames) receiver.handle(f);
      await expect(receiver.done).rejects.toThrow(/duplicate manifest/);
    }
  });

  it('settles immediately when the seed already holds the whole file', async () => {
    const manifest = await expectedManifest();
    const fp = await manifestFingerprint(manifest);
    const seed = nodeDigest();
    seed.update(new Uint8Array(20));
    const { err, back } = await runReceiver(manifest, false, {
      transferId: ID,
      manifestFingerprint: fp,
      files: new Map([[0, { haveBlocks: 3, seedDigest: seed }]]),
    });
    expect(err).toBeUndefined();
    expect(back.length).toBe(2); // resume_state, done
    const rs = back[0] as Extract<(typeof back)[number], { type: FrameType.ResumeState }>;
    expect(rs.files).toEqual([{ idx: 0, haveBlocks: 3 }]);
    expect(back[1]!.type).toBe(FrameType.Done);
  });
});

File: packages/wire/transfer_receiver_resume_test.go
package wire

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"strings"
	"testing"
)

// resumeManifest builds the manifest the senderFrames will stream, with the canonical
// fingerprint the seed must bind to (V13-PR06).
func resumeManifest(t *testing.T, id string, data []byte, blockSize int) (Manifest, string) {
	t.Helper()
	sum := sha256.Sum256(data)
	blocks := 0
	if len(data) > 0 {
		blocks = (len(data)-1)/blockSize + 1
	}
	raw := NewManifest([]FileEntry{{
		Idx: 0, Name: "f", Size: int64(len(data)), Mime: "application/octet-stream",
		LastModified: 1, BlockSize: blockSize, Blocks: blocks, FileDigest: hex.EncodeToString(sum[:]),
	}}, int64(len(data)))
	raw.TransferID = id
	validated, err := ValidateManifest(*raw)
	if err != nil {
		t.Fatal(err)
	}
	fp, err := ManifestFingerprint(validated)
	if err != nil {
		t.Fatal(err)
	}
	return validated, fp
}

// mustDigest feeds the given prefix into a fresh digest, mirroring a persisted high-water mark.
func mustDigest(t *testing.T, prefix []byte) Digest {
	t.Helper()
	d := NewSHA256Digest()
	d.Update(prefix)
	return d
}

// runReceiverWithSeed drives a receiver whose seed is the given claim and returns the Wait
// outcome plus the decoded back-channel frames.
func runReceiverWithSeed(t *testing.T, manifest Manifest, seed *ReceiverResume) (ReceiveResult, error, []ControlMsg) {
	t.Helper()
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	sf := newSenderFrames(t, keys)
	sf.ctrl(&manifest)
	if manifest.Files[0].Blocks > 0 {
		// Stream the whole file so a fresh receive completes; a resumed one ignores the held prefix.
		data := make([]byte, manifest.Files[0].Size)
		sf.blockData(data, manifest.Files[0].BlockSize, manifest.Files[0].BlockSize)
		sf.ctrl(NewComplete(CompletionDigest(manifest.Files)))
	}
	sink := &MemorySink{}
	var back outbox
	opts := ReceiverOptions{
		Send: back.push, SendDir: keys.J2O, RecvDir: keys.O2J, Sink: sink,
	}
	if seed != nil {
		opts.Resume = seed
	}
	r := NewReceiver(opts)
	for _, f := range sf.frames {
		r.Handle(f)
	}
	res, waitErr := r.Wait(context.Background())

	var msgs []ControlMsg
	for i, f := range back.snapshot() {
		o, err := Open(keys.J2O, uint64(i), f)
		if err != nil {
			t.Fatalf("open back frame %d: %v", i, err)
		}
		m, err := DecodeControl(o.Plaintext)
		if err != nil {
			t.Fatalf("decode back frame %d: %v", i, err)
		}
		msgs = append(msgs, m)
	}
	return res, waitErr, msgs
}

// TestReceiverRejectsBadResumeSeeds: a host-provided seed is a claim, not a trust anchor.
// Every impossible claim fails the receive closed before any of it is advertised.
func TestReceiverRejectsBadResumeSeeds(t *testing.T) {
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	_ = keys
	data := make([]byte, 20)
	manifest, fp := resumeManifest(t, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", data, 8)

	seed := func(mutate func(*ReceiverResume)) *ReceiverResume {
		s := &ReceiverResume{
			TransferID:          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
			ManifestFingerprint: fp,
			Files:               map[int]ResumeFileProgress{0: {HaveBlocks: 0}},
		}
		mutate(s)
		return s
	}

	cases := []struct {
		name string
		rs   *ReceiverResume
		want string
	}{
		{"uppercase fingerprint", seed(func(s *ReceiverResume) { s.ManifestFingerprint = strings.ToUpper(s.ManifestFingerprint) }), "resume seed manifestFingerprint must be 64 lowercase hex characters"},
		{"short fingerprint", seed(func(s *ReceiverResume) { s.ManifestFingerprint = fp[:10] }), "resume seed manifestFingerprint must be 64 lowercase hex characters"},
		{"wrong manifest fingerprint", seed(func(s *ReceiverResume) { s.ManifestFingerprint = strings.Repeat("0", 64) }), "resume seed manifest fingerprint does not match the authenticated manifest"},
		{"empty file set", seed(func(s *ReceiverResume) { s.Files = map[int]ResumeFileProgress{} }), "resume seed covers 0 files, manifest has 1"},
		{"missing file entry", seed(func(s *ReceiverResume) { s.Files = map[int]ResumeFileProgress{} }), "resume seed covers 0 files, manifest has 1"},
		{"haveBlocks above blocks", seed(func(s *ReceiverResume) { s.Files = map[int]ResumeFileProgress{0: {HaveBlocks: 4}} }), "resume seed haveBlocks 4 out of range for file 0 (blocks 3)"},
		{"haveBlocks negative", seed(func(s *ReceiverResume) { s.Files = map[int]ResumeFileProgress{0: {HaveBlocks: -1}} }), "resume seed haveBlocks -1 out of range for file 0 (blocks 3)"},
	}
	for _, c := range cases {
		c := c
		t.Run(c.name, func(t *testing.T) {
			_, err, msgs := runReceiverWithSeed(t, manifest, c.rs)
			if err == nil || !strings.Contains(err.Error(), c.want) {
				t.Fatalf("Wait error = %v, want one mentioning %q", err, c.want)
			}
			// The failure must be a sink_error fail frame; no resume_state may ever be advertised.
			if len(msgs) != 1 || msgs[0].FrameType() != FrameFail {
				t.Fatalf("back-channel = %#v, want a single fail frame", msgs)
			}
		})
	}
}

// TestReceiverAdvertisesFingerprintBoundResumeState: a valid seed is advertised only bound to
// the authenticated manifest's canonical fingerprint, and only the missing suffix is streamed.
func TestReceiverAdvertisesFingerprintBoundResumeState(t *testing.T) {
	data := make([]byte, 20)
	manifest, fp := resumeManifest(t, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", data, 8)

	prefix := 2 * 8
	seedDigest := NewSHA256Digest()
	seedDigest.Update(data[:prefix])
	seed := &ReceiverResume{
		TransferID:          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
		ManifestFingerprint: fp,
		Files:               map[int]ResumeFileProgress{0: {HaveBlocks: 2, SeedDigest: seedDigest}},
	}
	_, err, msgs := runReceiverWithSeed(t, manifest, seed)
	if err != nil {
		t.Fatalf("Wait: %v", err)
	}
	// resume_state, then an ack for every block the test streamed (the held 0-1 are acked
	// from the resumed high-water mark; block 2 is verified normally), then done.
	if len(msgs) != 5 {
		t.Fatalf("back-channel has %d frames, want 5 (resume_state, ack ×3, done)", len(msgs))
	}
	rs, ok := msgs[0].(*ResumeState)
	if !ok {
		t.Fatalf("first back frame = %T, want resume_state", msgs[0])
	}
	if rs.ManifestFingerprint != fp {
		t.Fatalf("advertised fingerprint = %q, want %q", rs.ManifestFingerprint, fp)
	}
	if len(rs.Files) != 1 || rs.Files[0].HaveBlocks != 2 {
		t.Fatalf("advertised resume_state = %#v, want haveBlocks 2", rs)
	}
	for i := 1; i <= 3; i++ {
		ack, ok := msgs[i].(*Ack)
		if !ok || ack.BlockIdx != i-1 {
			t.Fatalf("back frame %d = %#v, want ack for block %d", i, msgs[i], i-1)
		}
	}
	if msgs[4].FrameType() != FrameDone {
		t.Fatalf("last back frame = %T, want done", msgs[4])
	}
}

// TestReceiverIgnoresSeedForDifferentTransfer: a seed bound to another transfer is a different
// receive; the receiver starts fresh and advertises an all-zero resume_state for this one.
func TestReceiverIgnoresSeedForDifferentTransfer(t *testing.T) {
	data := make([]byte, 20)
	manifest, _ := resumeManifest(t, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", data, 8)
	staleSeed := NewSHA256Digest()
	staleSeed.Update(make([]byte, 8))
	seed := &ReceiverResume{
		TransferID:          "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
		ManifestFingerprint: strings.Repeat("0", 64),
		Files:               map[int]ResumeFileProgress{0: {HaveBlocks: 2, SeedDigest: staleSeed}},
	}
	_, err, msgs := runReceiverWithSeed(t, manifest, seed)
	if err != nil {
		t.Fatalf("Wait: %v", err)
	}
	rs, ok := msgs[0].(*ResumeState)
	if !ok {
		t.Fatalf("first back frame = %T, want resume_state", msgs[0])
	}
	if len(rs.Files) != 1 || rs.Files[0].HaveBlocks != 0 {
		t.Fatalf("advertised resume_state = %#v, want an all-zero high-water mark", rs)
	}
}

// TestReceiverReanswersIdenticalResumeStateOnDuplicateManifest: a cutover can deliver the
// manifest twice; the identical duplicate is re-answered with the exact same resume_state
// (never a re-validated one), while a different manifest is a protocol violation.
func TestReceiverReanswersIdenticalResumeStateOnDuplicateManifest(t *testing.T) {
	data := make([]byte, 20)
	manifest, fp := resumeManifest(t, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", data, 8)
	seed := &ReceiverResume{
		TransferID:          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
		ManifestFingerprint: fp,
		Files: map[int]ResumeFileProgress{
			0: {HaveBlocks: 1, SeedDigest: mustDigest(t, data[:8])},
		},
	}
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	sf := newSenderFrames(t, keys)
	sf.ctrl(&manifest)
	sf.ctrl(&manifest)
	sf.blockData(data, 8, 8)
	sf.ctrl(NewComplete(CompletionDigest(manifest.Files)))

	sink := &MemorySink{}
	var back outbox
	r := NewReceiver(ReceiverOptions{
		Send: back.push, SendDir: keys.J2O, RecvDir: keys.O2J, Sink: sink, Resume: seed,
	})
	for _, f := range sf.frames {
		r.Handle(f)
	}
	if _, err := r.Wait(context.Background()); err != nil {
		t.Fatalf("Wait: %v", err)
	}
	var msgs []ControlMsg
	for i, f := range back.snapshot() {
		o, err := Open(keys.J2O, uint64(i), f)
		if err != nil {
			t.Fatalf("open back frame %d: %v", i, err)
		}
		m, err := DecodeControl(o.Plaintext)
		if err != nil {
			t.Fatalf("decode back frame %d: %v", i, err)
		}
		msgs = append(msgs, m)
	}
	if len(msgs) != 6 {
		t.Fatalf("back-channel has %d frames, want 6 (resume_state ×2, ack ×3, done)", len(msgs))
	}
	first, ok1 := msgs[0].(*ResumeState)
	second, ok2 := msgs[1].(*ResumeState)
	if !ok1 || !ok2 || !resumeStatesEqual(first, second) {
		t.Fatalf("duplicate manifest answers = %#v / %#v, want identical resume_state messages", msgs[0], msgs[1])
	}
	if msgs[5].FrameType() != FrameDone {
		t.Fatalf("last back frame = %T, want done", msgs[5])
	}

	// A *different* manifest arriving after the first was applied is a violation: the seed
	// already validated against the first manifest, and the second is not that transfer.
	other, _ := resumeManifest(t, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", make([]byte, 40), 8)
	sf2 := newSenderFrames(t, keys)
	sf2.ctrl(&manifest)
	sf2.ctrl(&other)
	sink2 := &MemorySink{}
	var back2 outbox
	r2 := NewReceiver(ReceiverOptions{
		Send: back2.push, SendDir: keys.J2O, RecvDir: keys.O2J, Sink: sink2, Resume: seed,
	})
	for _, f := range sf2.frames {
		r2.Handle(f)
	}
	_, err = r2.Wait(context.Background())
	if err == nil || !strings.Contains(err.Error(), "duplicate manifest") {
		t.Fatalf("Wait error = %v, want duplicate manifest", err)
	}
}

// TestReceiverAllCompleteSeedSettles: a seed whose high-water mark already covers the whole
// file needs no data; the receiver advertises it and settles on the terminal Complete.
func TestReceiverAllCompleteSeedSettles(t *testing.T) {
	data := make([]byte, 20)
	manifest, fp := resumeManifest(t, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", data, 8)
	seedDigest := NewSHA256Digest()
	seedDigest.Update(data)
	seed := &ReceiverResume{
		TransferID:          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
		ManifestFingerprint: fp,
		Files:               map[int]ResumeFileProgress{0: {HaveBlocks: 3, SeedDigest: seedDigest}},
	}
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	sf := newSenderFrames(t, keys)
	sf.ctrl(&manifest)
	sf.ctrl(NewComplete(CompletionDigest(manifest.Files)))

	sink := &MemorySink{}
	var back outbox
	r := NewReceiver(ReceiverOptions{
		Send: back.push, SendDir: keys.J2O, RecvDir: keys.O2J, Sink: sink, Resume: seed,
	})
	for _, f := range sf.frames {
		r.Handle(f)
	}
	res, err := r.Wait(context.Background())
	if err != nil {
		t.Fatalf("Wait: %v", err)
	}
	if res.Digest != hexSHA256(data) {
		t.Fatalf("digest = %s, want %s", res.Digest, hexSHA256(data))
	}
	if !sink.IsClosed() {
		t.Fatal("sink not closed on done")
	}
	var msgs []ControlMsg
	for i, f := range back.snapshot() {
		o, err := Open(keys.J2O, uint64(i), f)
		if err != nil {
			t.Fatalf("open back frame %d: %v", i, err)
		}
		m, err := DecodeControl(o.Plaintext)
		if err != nil {
			t.Fatalf("decode back frame %d: %v", i, err)
		}
		msgs = append(msgs, m)
	}
	if len(msgs) != 2 {
		t.Fatalf("back-channel has %d frames, want 2 (resume_state, done)", len(msgs))
	}
	rs, ok := msgs[0].(*ResumeState)
	if !ok || rs.Files[0].HaveBlocks != 3 {
		t.Fatalf("first back frame = %#v, want resume_state with haveBlocks 3", msgs[0])
	}
	if msgs[1].FrameType() != FrameDone {
		t.Fatalf("second back frame = %T, want done", msgs[1])
	}
}
